### ISSUE TITLE
Issue when parsing a file with mixed EOL

### ISSUE THREAD

**Issue by maros** [Tux/Text-CSV_XS#59]
I have a slightly broken file with mixed EOL (\n, \r\n and only \r - according to https://metacpan.org/pod/Text::CSV_XS#eol this should be handled automatically) that causes the parts of the file to be skipped silently. No error is reported.

```perl
use 5.038;
use Text::CSV_XS;

my $csv = Text::CSV_XS->new ({
    sep_char => ';',
});

# Write test data
open my $fhw, '>','csvtest.csv';
print $fhw "Antelope;snort\r\n";
print $fhw "Badger;growl\n"; # only newline
print $fhw "Bat;screech\r\n";
print $fhw "Bear;roar\r"; # only carriage return - no newline
print $fhw "Bee;buzz\r\n";
print $fhw "Camel;grunt\r\n";
print $fhw "Crow;caw\n"; # only newline
print $fhw "Deer;bellow\r\n";
print $fhw "Dolphin;click\r\n";
$fhw->close;

# Read test data
open my $fhr, '<','csvtest.csv';
while (my $row = $csv->getline($fhr)) {
    say sprintf '%s makes %s',($row->[0]||'-'),($row->[1]||'-');
}

# Remove test data
unlink 'csvtest.csv';
```

the output of the given example is
```
Antelope makes snort
Badger makes growl
Bat makes screech
Bear makes roar
Bee makes buzz
- makes -
Camel makes grunt
- makes -
- makes -
- makes -
```
There is an empty row between 'Bee' and 'Camel' which shouldn't be there, as well as all rows after 'Camel' are being skipped.

Tested with Text::CSV_XS 1.56 and perl 5.40.0

**Comment by Tux** [comment]
This one is relatively easy to expain, and *maybe* requires additional documentation. I value your feedback on this.

What the docs say is that `\r`, `\r\n`, and `\n` are all valid, which still holds, but what happens inside is that the parser "remembers" the found EOL for efficiency. The workaround in your case is to make the parser "forget" what it found:

```
use 5.038;
use Text::CSV_XS;

my $csv = Text::CSV_XS->new ({
    sep_char  => ";",
    auto_diag => 1,
    });

my $tfn = "issue-59.csv";

# Write test data
open my $fh, ">", $tfn or die "$tfn: $!\n";
print   $fh "Antelope;snort\r\n";
print   $fh "Badger;growl\n"; # only newline
print   $fh "Bat;screech\r\n";
print   $fh "Bear;roar\r"; # only carriage return - no newline
print   $fh "Bee;buzz\r\n";
print   $fh "Camel;grunt\r\n";
print   $fh "Crow;caw\n"; # only newline
print   $fh "Deer;bellow\r\n";
print   $fh "Dolphin;click\r\n";
close   $fh;

# Read test data
open $fh, "<", $tfn or die "$tfn: $!\n";
while (my $row = $csv->getline ($fh)) {
    printf "%s makes %s\n", map { $_ || "-" } $row->[0], $row->[1];
    $csv->eol (undef); # Forget what EOL it found
    }

# Remove test data
unlink $tfn;
```
→
```
Antelope makes snort
Badger makes growl
Bat makes screech
Bear makes roar
Bee makes buzz
Camel makes grunt
Crow makes caw
Deer makes bellow
Dolphin makes click
```

Options might be
- document current behavior better, possibly with above example
- have a new option eol_nocache defaulting to false and possibly slowing down parsing when enabled
- make no-cache the default

I don't think that last option is viable, as it is backwards incompatible and might break existing scripts
I don't know what the impact (on speed) would for that new option

**Comment by Tux** [comment]
And sorry for the late reply. Been busy

**Comment by maros** [comment]
Thank you for pointing out this solution. From my point of view a note in the documentation is sufficient, since it may be easily implemented in the user code.

However I believe that this issue isn't fully adressed just by resetting the eol cache after parsing each row, since currently there is no way to reliably detect such broken files (apart from using the overly-strict strict flag or skip_empty_rows). I would really apreciate it if error_diag could report such cases.

And also sorry for the late reply - was on vaccation

**Comment by Tux** [comment]
I meanwhile thought of yet another option: `strict_eol`. With that implemented as die/croack/error (like with `strict`) when EOL is inconsistent, it would fix your issue?

Just a thought, and I have no idea yet on the speed-impact. Makes sense though

**Comment by maros** [comment]
Thank you for your change. 

However after testing the changes it seems that the "2016 - EOL - Inconsistent EOL" error is always called on the first record whenever strict_eol is set in conjunction with auto_diag


When i run
`
use 5.038;
use Text::CSV_XS;

my $csv = Text::CSV_XS->new ({
    sep_char => ';',
    auto_diag => 9,
    diag_verbose => 1,
    strict_eol => 1,
});

# Write test data
open my $fhw, '>','csvtest.csv'; 
print $fhw "Antelope;snort\r\n"; # always CR-NL
print $fhw "Badger;growl\r\n";
print $fhw "Bat;screech\r\n";
$fhw->close;

# Read test data
open my $fhr, '<','csvtest.csv';
while (my $row = $csv->getline($fhr)) {
    say sprintf '%s makes %s',($row->[0]||'-'),($row->[1]||'-');
}

# Remove test data
unlink 'csvtest.csv';
`

the output is
`
# CSV_XS ERROR: 2016 - EOL - Inconsistent EOL @ rec 0 pos 15 field 2
Antelope;snort

              ^
`

**Comment by Tux** [comment]
Lots has changed since. I also wrote tests. Changes have not been pushed, as there still is a conflict-situation I have to resolve.

In *theory*, would `strict_eol` meet your expectations?

**Comment by maros** [comment]
Yes ```strict_eol``` meets my usecase. (Apart form the previously reported case) it also behaved corretly when presented various broken files.

**Comment by Tux** [comment]
OK, that took much longer than anticipated. It is a snakepit with still room for improvement

Could you try https://github.com/Tux/Text-CSV_XS/commit/dd5ba28436e3ae0e4a24322c9954b61258138b31 ?
`strict_eol => 1` = warn only (and only once), `strict_eol => 2` is die on inconsistent EOL use

**Comment by maros** [comment]
Thank you for your fix. I found a minor issue with the strict_eol diag message: If an inconsitent EOL is detected, the error message contains parts of the next line.
 
 given this input data
 ```perl
print $fhw "Antelope;snort\r\n";
print $fhw "Badger;growl\r\n";
print $fhw "Bat;screech\r\n";
print $fhw "Bear;roarrrrrrrrrrr\r"; # only carriage return - no newline
print $fhw "Bee;buzz\r\n";
```

i get the following error message with the above test script
```
Antelope makes snort
Badger makes growl
Bat makes screech
# CSV_XS ERROR: 2016 - EOL - Inconsistent EOL @ rec 3 pos 21 field 2 (XS#2146)
Bee;buzzrrrrrrrrrrr
```

not sure if rec should be 3 or 4, but the error input line should be "Bear;roarrrrrrrrrrr" not "Bee;buzzrrrrrrrrrrr"

**Comment by Tux** [comment]
I do not yet know (for sure) what the error line cntent should be as there are two ways to look at this (new) error: is it a "wrong" way to end the previous record or a wrong way to start the next.

Main reason is to detect inconsistencies, and that has been achieved.
I want to check what to do with a few edge cases of which I currently think the result isn't DWIM.
I also want to look at supporting more in/out combinations of `csv ()`

Happy X-mas and a sane new-year.
Thanks for the feedback

**Comment by maros** [comment]
oh sorry i just didn't realize that the actual output from the error message actually was "Bear;roarrrrrrrrrrr\rBee;buzz\r\n" (which is totally fine) - i just read the resulting "Bee;buzzrrrrrrrrrrr" from the command line output and thought that this was the actual output from Text::CSV_XS.

Yes, i can confirm that the module now works as expected. Thank you very much for your effort - and also happy holidays to you
